File: CompositeEntity/CompositeEntity.js
/**
 * Created by jamesyan on 2017/3/25.
 */

var DependentObject1=function(){
    this.setData=function(data){
        this.data=data;
    }
    this.getData=function(){
        return this.data
    }
}

var DependentObject2=function(){
    this.setData=function(data){
        this.data=data;
    }
    this.getData=function(){
        return this.data
    }
}

var CoarseGrainedObject=function(){
    this.do1=new DependentObject1();
    this.do2=new DependentObject2();
    this.setData=function(data1,data2){
        this.do1.setData(data1);
        this.do2.setData(data2);
    }

    this.getData=function(){
        return [this.do1.getData(),this.do2.getData()];
    }
}

var CompositeEntity=function(){
    this.cgo=new CoarseGrainedObject();
    this.getData=function(){
        return this.cgo.getData();
    }
    this.setData=function(data1,data2){
        this.cgo.setData(data1,data2);
    }
}

var Client=function(){
    this.entity=new CompositeEntity();
    this.printData=function(){
        for(var d in this.entity.getData()){
            var data=this.entity.getData()[d];
            console.log("Data :"+data);
        }
    }
    this.setData=function(data1,data2){
        this.entity.setData(data1,data2);
    }

}

File: CompositeEntity/CompositeEntity.go
package main

import "fmt"

type DependentObject1 struct {
	data string
}

func(d *DependentObject1)setData(data string){
	d.data=data
}

func(d *DependentObject1)getData() string{
	return d.data
}

type DependentObject2 struct {
	data string
}

func(d *DependentObject2)setData(data string){
	d.data=data
}

func(d *DependentObject2)getData() string{
	return d.data
}


type CoarseGrainedObject struct {
	do1 *DependentObject1
	do2 *DependentObject2
}

func(c *CoarseGrainedObject)setData(data1,data2 string){
	c.do1.setData(data1)
	c.do2.setData(data2)
}

func(c *CoarseGrainedObject)getData()[]string{
	return []string{c.do1.getData(),c.do2.getData()}
}

type CompositeEntity struct {
	cgo *CoarseGrainedObject
}

func(c *CompositeEntity)setData(data1,data2 string){
	c.cgo.setData(data1,data2)
}

func(c *CompositeEntity)getData()[]string{
	return c.cgo.getData()
}

type Client struct {
    compositeEntity *CompositeEntity
}

func(c *Client)printData(){
	for _,composite:=range c.compositeEntity.getData(){
		fmt.Println("Data:",composite)
	}
}


func(c *Client)setData(data1,data2 string){
	c.compositeEntity.setData(data1,data2)
}


func main(){
	client:=&Client{&CompositeEntity{&CoarseGrainedObject{new(DependentObject1),new(DependentObject2)}}}
	client.setData("Test", "Data")
	client.printData()

	client.setData("secend Test", "Data1")
	client.printData()

}



